packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,285 @@
+we still own that amount.
+*/
+export function addressTypeFactor(
+  transactions: Array<Transaction>,
+  walletAddressType: string,
+): number {
+  let P2WSH: number = 0;
+  let P2PKH: number = 0;
+  let P2SH: number = 0;
+  let atf: number = 1;
+  transactions.forEach((tx) => {
+    tx.vout.forEach((output) => {
+      let address = output.scriptPubkeyAddress;
+      if (address.startsWith("bc")) {
+        //Bech 32 Native Segwit (P2WSH) or Taproot
+        P2WSH += 1;
+      } else if (address.startsWith("1")) {
+        // Legacy (P2PKH)
+        P2PKH += 1;
+      } else {
+        // Segwith (P2SH)
+        P2SH += 1;
+      }
+    });
+  });
+
+  if (walletAddressType == "P2WSH" && P2WSH != 0 && (P2SH != 0 || P2PKH != 0)) {
+    atf = 1 / (P2WSH + 1);
+  } else if (
+    walletAddressType == "P2PKH" &&
+    P2PKH != 0 &&
+    (P2SH != 0 || P2WSH != 0)
+  ) {
+    atf = 1 / (P2PKH + 1);
+  } else if (
+    walletAddressType == "P2SH" &&
+    P2SH != 0 &&
+    (P2WSH != 0 || P2PKH != 0)
+  ) {
+    atf = 1 / (P2SH + 1);
+  } else {
+    atf = 1;
+  }
+  return atf;
+}
+
+/* 
+The spread factor using standard deviation helps in assessing the dispersion of UTXO values.
+In Bitcoin privacy, spreading UTXOs reduces traceability by making it harder for adversaries
+to link transactions and deduce the ownership and spending patterns of users.
+*/
+export function utxoSpreadFactor(utxos: Array<WalletUTXOs>): number {
+  const amounts: Array<number> = utxos.map((utxo) =>
+    utxo.utxos.reduce((sum, utxo) => sum + utxo.value, 0),
+  );
+  const mean: number =
+    amounts.reduce((sum, amount) => sum + amount, 0) / amounts.length;
+  const variance: number =
+    amounts.reduce((sum, amount) => sum + Math.pow(amount - mean, 2), 0) /
+    amounts.length;
+  const stdDev: number = Math.sqrt(variance);
+  return stdDev / (stdDev + 1);
+}
+
+/* 
+The weightage is ad-hoc to normalize the privacy score based on the number of UTXOs in the set.
+- 0 for UTXO set length >= 50
+- 0.25 for UTXO set length >= 25 and <= 49
+- 0.5 for UTXO set length >= 15 and <= 24
+- 0.75 for UTXO set length >= 5 and <= 14
+- 1 for UTXO set length < 5
+*/
+export function utxoSetLengthWeight(utxos: Array<WalletUTXOs>): number {
+  let utxoSetLength = utxos.reduce((sum, utxo) => sum + utxo.utxos.length, 0);
+  let weight: number;
+  if (utxoSetLength >= 50) {
+    weight = 0;
+  } else if (utxoSetLength >= 25 && utxoSetLength <= 49) {
+    weight = 0.25;
+  } else if (utxoSetLength >= 15 && utxoSetLength <= 24) {
+    weight = 0.5;
+  } else if (utxoSetLength >= 5 && utxoSetLength <= 14) {
+    weight = 0.75;
+  } else {
+    weight = 1;
+  }
+  return weight;
+}
+
+/*
+UTXO Value Weightage Factor is a combination of UTXO Spread Factor and UTXO Set Length Weight.
+It signifies the combined effect of how well spreaded the UTXO Set is and how many number of UTXOs are there.
+*/
+export function utxoValueWeightageFactor(utxos: Array<WalletUTXOs>): number {
```

Comment:
```suggestion
export function utxoValueWeightageFactor(utxos: WalletUTXOs[]): number {
```
in fact, the pluralization is a little weird I think too (I think it goes to my question above about the type?). 

It looks like this is an object that's associating an address with outputs that have been spent to it? Maybe a cleaner way to do this would be to have a separate type that maps addresses to their utxos. which would be typed something like:

```
interface AddressUtxos {
    [address:string]: UTXO[]
}
```